src/equations/shallow_water_two_layer_1d.jl

```diff
@@ -0,0 +1,586 @@
+bottom topography function ``b(x)``. The conservative variable water height ``h_2`` is measured 
+from the bottom topography ``b`` and ``h_1`` relative to ``h_2``, therefore one also defines the 
+total water heights as ``H_1 = h_1 + h_2 + b`` and ``H_2 = h_2 + b``.
+
+The additional quantity ``H_0`` is also available to store a reference value for the total water
+height that is useful to set initial conditions or test the "lake-at-rest" well-balancedness.
+
+The bottom topography function ``b(x)`` is set inside the initial condition routine
+for a particular problem setup.
+
+In addition to the unknowns, Trixi currently stores the bottom topography values at the 
+approximation points despite being fixed in time. This is done for convenience of computing the 
+bottom topography gradients on the fly during the approximation as well as computing auxiliary 
+quantities like the total water height ``H`` or the entropy variables.
+This affects the implementation and use of these equations in various ways:
+* The flux values corresponding to the bottom topography must be zero.
+* The bottom topography values must be included when defining initial conditions, boundary 
+  conditions or source terms.
+* [`AnalysisCallback`](@ref) analyzes this variable.
+* Trixi's visualization tools will visualize the bottom topography by default.
+
+A good introduction for the 2L-SWE is available in Chapter 12 of the book:
+- Benoit Cushman-Roisin (2011)
+  Introduction to geophyiscal fluid dynamics: physical and numerical aspects
+  https://www.sciencedirect.com/bookseries/international-geophysics/vol/101/suppl/C
+  ISBN: 978-0-12-088759-0
+"""
+struct ShallowWaterTwoLayerEquations1D{RealT<:Real} <: AbstractShallowWaterEquations{1,5}
+  gravity::RealT # gravitational constant
+  H0::RealT      # constant "lake-at-rest" total water height
+  rho1::RealT    # lower layer density
+  rho2::RealT    # upper layer density
+  r::RealT       # ratio of rho1 / rho2
+end
+
+# Allow for flexibility to set the gravitational constant within an elixir depending on the
+# application where `gravity_constant=1.0` or `gravity_constant=9.81` are common values.
+# The reference total water height H0 defaults to 0.0 but is used for the "lake-at-rest"
+# well-balancedness test cases. Densities must be specificed such that rho_1 <= rho_2.
+function ShallowWaterTwoLayerEquations1D(; gravity_constant, H0=0.0, rho1, rho2)
+  # Assign density ratio if rho1 <= rho_2
+  if rho1 / rho2 > 1
+    error("Invalid input: Densities must be chosen such that rho1 <= rho2")
+  else
+    r = rho1 / rho2
+  end
+  ShallowWaterTwoLayerEquations1D(gravity_constant, H0, rho1, rho2, r)
+end
+
+have_nonconservative_terms(::ShallowWaterTwoLayerEquations1D) = True()
+varnames(::typeof(cons2cons), ::ShallowWaterTwoLayerEquations1D) = ("h1", "h1_v1",
+                                                                    "h2", "h2_v2", "b")
```

Comment:
I have used the following notation in 2D:
`v`: x-velocity
`w`: y-velocity
`1`: upper layer
`2`: lower layer

Reply:
Whew. Unless the `v`/`w`/`1`/`2` notation is super consistently used everywhere that SWE appear, I'd go for something that is consistent with our current notation. Why not use something like
```
h_upper, h_upper_v1, h_lower, h_lower_v1, b
```
? That would be self-descriptive in the upper/lower part and consistent with the rest of the equations in the `v1` part.

There might be better alternatives than my suggestion, but in the current approach I really see how easy it would be to confuse `v1` with the x-velocity and `w` with the z-velocity, since this is how they are commonly used.

Reply:
I definitely agree with that!

Reply:
The `1`/`2` notation is pretty much standard (I've only found a single reference where a upper / lower notation is used).  A benefit is that if at some point, we'd like to extend the model to multiple layers one can just add `h3`,`h4` and so on.
Currently this is not planned so I could use upper, lower. However, expressions will get super long as we have to distinguish between layer velocities and directions. So this line
` h1, h1_v1, h1_w1, h2, h2_v2, h2_w2, _ = u`
would become
`h_upper, h_upper_v_upper, h_upper_w_upper, h_lower, h_lower_v_lower, h_lower_w_lower, _ = u`,
which I find really hard to read.

Reply:
I don't understand why one  needs the double `upper`. Is there ever a case of `h_lower_v_upper`?

I see your point about multiple layers. But if you are to be flexible with the number of layers, you will not be able to use fixed variable names anymore anyways, or am I missing something? For instance:
https://github.com/trixi-framework/Trixi.jl/blob/06635139bf88808601552d751c642563ae8faa76/src/equations/compressible_euler_multicomponent_2d.jl#L92-L97

Reply:
No, there are no mixed variables, I used this notation because it shows directly which layer `h`, and `v` belong to and that they differ between layers. But if we are using this longer notation I also think it is better to just state the layer once. What do you think about this suggestion?
`h_upper, h_v1_upper, h_v2_upper, h_lower, h_v1_lower, h_v2_lower, _ = u`

I didn't really look into how it would be implemented with multiple layers. I guess at least for the output you need to assign numbers to it, so you would have
`h_1, h_v1_1, h_v1_2, h_2, h_v1_2, h_v2_2, h_3, h_v1_3, h_v2_3`?

Reply:
I am okay with the option to use `h_upper, h_v1_upper, h_v2_upper, h_lower, h_v1_lower, h_v2_lower, _ = u`. We can clarify in the docstring the meaning of upper and lower.

Reply:
I changed the variable names to the `h_upper, h_v1_upper, h_v2_upper, h_lower, h_v1_lower, h_v2_lower, _ = u` notation in commit 20eafbfb2ba8635a90f6dc4f1f8211aca566b515